passes/opt/opt_expr.cc

```diff
@@ -1251,7 +1251,145 @@ void replace_const_cells(RTLIL::Design *design, RTLIL::Module *module, bool cons
 					}
 			}
 		}
+		if (do_fine && (cell->type.in("$eq", "$ne")))
+		{
+			RTLIL::SigSpec a_sig, b_sig, y_sig;
+			
+			int a_width, b_width;
+			bool a_signed, b_signed;
+			
+			a_sig = assign_map(cell->getPort("\\A"));
+			b_sig = assign_map(cell->getPort("\\B"));
+			y_sig = assign_map(cell->getPort("\\Y"));
+			a_width = cell->parameters["\\A_WIDTH"].as_int();
+			b_width = cell->parameters["\\B_WIDTH"].as_int();
+			a_signed = cell->parameters["\\A_SIGNED"].as_bool();
+			b_signed = cell->parameters["\\B_SIGNED"].as_bool();
+			
+			dict<SigBit, RTLIL::SigSpec> compared_by_const;
+			pool<SigBit> signals_map;
+			SigBit polarity_bit = cell->type == "$eq" ? State::S0 : State::S1;
+
+			if (a_signed || b_signed)
+				goto next_cell;
+
+			for (int i = 0; i < a_width; i++)
+			{
+				if ((a_sig[i] ==  RTLIL::State::S0) || (a_sig[i] ==  RTLIL::State::S1))
+				{
+					if (compared_by_const.count(b_sig[i]))
+					{
+						if (compared_by_const.at(b_sig[i]) != a_sig[i])
+						{
+							// case where comparing signal against S0 and S1
+							log("Replacing cell `%s' in module `%s' with zero-driver.\n",
+								cell->name.c_str(), module->name.c_str());
+							
+							module->connect(RTLIL::SigSig(y_sig, RTLIL::SigSpec(polarity_bit, y_sig.size())));
+							module->remove(cell);
+							
+							did_something = true;
+							goto next_cell;
+						}
+					} else
+					{
+						compared_by_const[b_sig[i]] = a_sig[i];
+					}
+				}
+			}
+			for (int i = 0; i < b_width; i++)
+			{
+				if ((b_sig[i] ==  RTLIL::State::S0) || (b_sig[i] ==  RTLIL::State::S1))
+				{
+					if (compared_by_const.count(a_sig[i]))
+					{
+						if (compared_by_const.at(a_sig[i]) != b_sig[i])
+						{
+							// case where comparing signal against S0 and S1
+							log("Replacing cell `%s' in module `%s' with zero-driver.\n",
+								cell->name.c_str(), module->name.c_str());
+							
+							module->connect(RTLIL::SigSig(y_sig, RTLIL::SigSpec(polarity_bit, y_sig.size())));
+							module->remove(cell);
+							
+							did_something = true;
+							goto next_cell;
+						}
+					} else
+					{
+						compared_by_const[b_sig[i]] = b_sig[i];
+					}
+				}
+			}
 
+			if (a_width + b_width > 2)
+			{
+				for (int i = 0; i < a_width; i++)
+				{
+					signals_map.insert(a_sig[i]);
+				}
+				for (int i = 0; i < b_width; i++)
+				{
+					signals_map.insert(b_sig[i]);
+				}
+				
+				RTLIL::SigSpec c_sig, sig_const, ss;
+				if (signals_map.size() == 2)
+				{
+					for (auto ss: signals_map)
+					{
+						if ((ss == RTLIL::State::S0) || (ss == RTLIL::State::S1))
+						{
+							sig_const = ss;
+						} else
+						{
+							c_sig = ss;
+						}
+					}
```

Comment:
Imagine ({A,B} = {B,A}) . It has (a_width + b_width > 2) and (signals_map.size() == 2) but sig_const will be used uninitialized in the condition below.

Reply:
Added guard for doing this optimisation only for cells that has no x's or z's on inputs.